[Bug] Fixed primal weather interaction.

Arxalc opened the pull request:
<!-- Make sure the title includes categorization (i.e. [Bug], [QoL], [Localization]) -->
<!-- Make sure that this PR is not overlapping with someone else's work -->
<!-- Please try to keep the PR self-contained (and small) -->

## What are the changes?
<!-- Summarize what are the changes from a user perspective on the application -->
Closes #1715
* Primal weather can overwrite each other
* Primal weather ends upon switching out or fainting

## Why am I doing these changes?
<!-- Explain why you decided to introduce these changes -->
<!-- Does it come from an issue or another PR? Please link it -->
<!-- Explain why you believe this can enhance user experience -->
Interactions weren't working before.

## What did change?
<!-- Explicitly state what are the changes introduced by the PR -->
<!-- You can make use of a comparison between what was the state before and after your PR changes -->
If a primal weather was already in effect, it could not be overriden by another -> now it can
Primal weather would not revert upon Pokemon switching out or fainting -> now it does

### Screenshots/Videos
<!-- If your change is changing anything on the user experience, please provide visual proofs of it -->
<!-- Please take screenshots/videos before and after your changes, to show what is brought by this PR -->

https://github.com/pagefaultgames/pokerogue/assets/63990624/68eb7021-59a7-48ea-802a-5692f72bb95c


## How to test the changes?
<!-- How can a reviewer test your changes once they check out on your branch? -->
<!-- Did you just make use of the `src/overrides.ts` file? -->
<!-- Did you introduce any automated tests? -->
<!-- Do the reviewer need to do something special in order to test your change? -->
Override ability to Desolate Land, Primordial Sea, or Delta Stream.

## Checklist
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I tested the changes (manually)?
    - [x] Are all unit tests still passing? (`npm run test`)
- [x] Are the changes visual?
  - [x] Have I provided screenshots/videos of the changes?

Tempo-anon commented on src/data/ability.ts:
```diff
@@ -1728,7 +1728,9 @@ export class PostSummonWeatherChangeAbAttr extends PostSummonAbAttr {
 
   applyPostSummon(pokemon: Pokemon, passive: boolean, args: any[]): boolean {
-    if (!pokemon.scene.arena.weather?.isImmutable()) {
+    if ((this.weatherType as WeatherType === WeatherType.HEAVY_RAIN || 
+      this.weatherType as WeatherType === WeatherType.HARSH_SUN || 
+      this.weatherType === WeatherType.STRONG_WINDS) || !pokemon.scene.arena.weather?.isImmutable()) {
```
Wonder if it would be worth having a util function in weather.ts `isPrimalWeather`

Tempo-anon commented on src/data/ability.ts:
```diff
@@ -1856,6 +1858,16 @@ export class PreSwitchOutResetStatusAbAttr extends PreSwitchOutAbAttr {
 }
 
+/**
+ * Clears weather upon Primal Groudon/Kyogre/Rayquaza switch out
```
nit: this ability attribute is not exclusive to those Pokemon, but rather on the abilites heavy rain, harsh sun, and strong winds

Tempo-anon commented on src/data/ability.ts:
```diff
@@ -1856,6 +1858,16 @@ export class PreSwitchOutResetStatusAbAttr extends PreSwitchOutAbAttr {
+/**
+ * Clears weather upon Primal Groudon/Kyogre/Rayquaza switch out
+ */
+export class PreSwitchOutClearWeatherAttr extends PreSwitchOutAbAttr {
```
nit: `PreSwitchOutClearWeatherAbAttr`

Tempo-anon commented on src/data/ability.ts:
```diff
@@ -1856,6 +1858,16 @@ export class PreSwitchOutResetStatusAbAttr extends PreSwitchOutAbAttr {
+ */
+export class PreSwitchOutClearWeatherAttr extends PreSwitchOutAbAttr {
+  applyPreSwitchOut(pokemon: Pokemon, passive: boolean, args: any[]): boolean | Promise<boolean> {
+    pokemon.scene.arena.trySetWeather(WeatherType.NONE, false);
```
Won't this cause a bug if there are two Primal Groudons out and one of them switches out? The harsh sun should still be in effect

Arxalc replied:
You're right. Will try and fix.
Tested interactions on Showdown and found it works like this:
* Groudon vs. Kyogre. When Kyogre switches in, Desolate Land is overriden by Primordial Sea. If Kyogre dies, weather clears.
* Groudon 1 vs. Groudon 2. When Groudon 2 switches in, nothing special happens. If Groudon 1 dies, weather persists because Groudon 2 has the ability and is still on the field.

Tempo-anon replied:
What happens if Kyogre switches out after overwriting Desolate Land with Primordial Sea? If the weather still clears up it just means you need to add a check so on faint/switchout it checks the field to make sure no other Pokemon has that ability which means this ability attribute also needs to know what ability it's checking

Tempo-anon commented on src/data/ability.ts:
```diff
@@ -2982,6 +2994,20 @@ export class PostFaintAbAttr extends AbAttr {
+  }
+
+  applyPostFaint(pokemon: Pokemon, passive: boolean, attacker: Pokemon, move: PokemonMove, hitResult: HitResult, args: any[]): boolean {
+    pokemon.scene.arena.trySetWeather(WeatherType.NONE, false);
```
Same issue as above

Tempo-anon reviewed (changes requested):
There's some edge cases to consider here

Arxalc commented:
I added checking abilities for both switch out/faint so now all cases should be covered.

Fixed 1st comment case: Harsh sun persists after Desolate Land user faints and another Desolate Land ability remains
https://github.com/pagefaultgames/pokerogue/assets/63990624/277bb5d6-2ff2-4261-8f8c-dc666f80f546

2nd comment case: According to Showdown after overwriting a primal weather, weather should clear if Kyogre switches out/faints. This is intentional. 
https://github.com/pagefaultgames/pokerogue/assets/63990624/b6859ec9-34f5-46ea-bb22-415691a3468e

Tempo-anon commented on src/data/ability.ts:
```diff
@@ -1728,7 +1728,9 @@ export class PostSummonWeatherChangeAbAttr extends PostSummonAbAttr {
-    if (!pokemon.scene.arena.weather?.isImmutable()) {
+    if ((this.weatherType as WeatherType === WeatherType.HEAVY_RAIN ||
+      this.weatherType as WeatherType === WeatherType.HARSH_SUN ||
+      this.weatherType === WeatherType.STRONG_WINDS) || !pokemon.scene.arena.weather?.isImmutable()) {
```
nit: how come the strong winds check does not `as`?

Tempo-anon commented on src/data/ability.ts:
```diff
@@ -1856,6 +1858,29 @@ export class PreSwitchOutResetStatusAbAttr extends PreSwitchOutAbAttr {
+ * Clears Desolate Land/Primordial Sea/Delta Stream upon the Pokemon switching out.
+ */
+export class PreSwitchOutClearWeatherAbAttr extends PreSwitchOutAbAttr {
+  applyPreSwitchOut(pokemon: Pokemon, passive: boolean, args: any[]): boolean | Promise<boolean> {
```
Please add tsdoc to the `applyPreSwitchOut` function, especially since you're using `pokemon` in it

Tempo-anon commented on src/data/ability.ts:
```diff
@@ -1856,6 +1858,29 @@ export class PreSwitchOutResetStatusAbAttr extends PreSwitchOutAbAttr {
+ */
+export class PreSwitchOutClearWeatherAbAttr extends PreSwitchOutAbAttr {
+  applyPreSwitchOut(pokemon: Pokemon, passive: boolean, args: any[]): boolean | Promise<boolean> {
+    const currentAbility = pokemon.getAbility().id;
```
I believe `getAbility` does not take into account passives which might be worth considering

Tempo-anon commented on src/data/ability.ts:
```diff
@@ -1856,6 +1858,29 @@ export class PreSwitchOutResetStatusAbAttr extends PreSwitchOutAbAttr {
+    // Clear weather if ability matches weather and if no one else has the ability
+    if ((currentAbility === Abilities.DESOLATE_LAND && weatherType === WeatherType.HARSH_SUN) ||
+    (currentAbility === Abilities.PRIMORDIAL_SEA && weatherType === WeatherType.HEAVY_RAIN) ||
+    (currentAbility === Abilities.DESOLATE_LAND && weatherType === WeatherType.HARSH_SUN)) {
```
There's two desolate land/harsh sun checks here but no strong winds check

Tempo-anon commented on src/data/ability.ts:
```diff
@@ -1856,6 +1858,29 @@ export class PreSwitchOutResetStatusAbAttr extends PreSwitchOutAbAttr {
+    (currentAbility === Abilities.DESOLATE_LAND && weatherType === WeatherType.HARSH_SUN)) {
+
+      // Besides the user, check if another has the ability on the field.
+      if (pokemon.scene.getField(true).filter(p => p !== pokemon).filter(p => p.hasAbility(currentAbility)).length === 0) {
```
I wonder if there's a more elegant way of doing this

Tempo-anon commented on src/data/ability.ts:
```diff
@@ -2982,6 +3007,29 @@ export class PostFaintAbAttr extends AbAttr {
+    // Clear weather if ability matches weather and if no one else has the ability
+    if ((currentAbility === Abilities.DESOLATE_LAND && weatherType === WeatherType.HARSH_SUN) ||
+    (currentAbility === Abilities.PRIMORDIAL_SEA && weatherType === WeatherType.HEAVY_RAIN) ||
+    (currentAbility === Abilities.DESOLATE_LAND && weatherType === WeatherType.HARSH_SUN)) {
```
two desolate land checks here, which I guess makes sense since I assume this is copy/pasted. All my other comments about the other class also apply here

Tempo-anon reviewed (changes requested):
See some errors but looking good!

Arxalc commented:
1. Removed the `as` part. (tldr: I only know Java. Tried `==`. Had errors. Read you needed `as`. Turns out I just needed to change to `===` instead.)
2. Added passive abilities into consideration. (Balance-wise, unlikely to ever happen
3. I completely missed strong winds 🤡
4. Lmk if there's a better way to do this. I don't think it's worth adding util functions for something that'll only happen twice ever.

Tempo-anon commented on src/data/ability.ts:
```diff
@@ -2982,6 +3017,42 @@ export class PostFaintAbAttr extends AbAttr {
+   * @param args N/A
+   * @returns {boolean} Returns true if the weather clears, otherwise false.
+   */
+  applyPostFaint(pokemon: Pokemon, passive: boolean, attacker: Pokemon, move: PokemonMove, hitResult: HitResult, args: any[]): boolean {
```
Since we can make the assumption that there's only ever one weather at a time, would something like this make sense?
```ts
    let turnOffWeather = false;
    switch (weatherType) {
      case (WeatherType.HARSH_SUN):
        if (pokemon.hasAbility(Abilities.DESOLATE_LAND) 
          && pokemon.scene.getField(true).filter(p => p.hasAbility(Abilities.DESOLATE_LAND).length === 0)) {
            turnOffWeather = true;
        }
        break;
      case (WeatherType.HEAVY_RAIN):
        if (pokemon.hasAbility(Abilities.PRIMORDIAL_SEA) 
          && pokemon.scene.getField(true).filter(p => p.hasAbility(Abilities.PRIMORDIAL_SEA).length === 0)) {
            turnOffWeather = true;
        }
        break;
      case (WeatherType.STRONG_WINDS):
        if (pokemon.hasAbility(Abilities.DELTA_STREAM) 
          && pokemon.scene.getField(true).filter(p => p.hasAbility(Abilities.DELTA_STREAM).length === 0)) {
            turnOffWeather = true;
        }
        break;
      default:
    }
    if (turnOffWeather) {
      pokemon.scene.arena.trySetWeather(WeatherType.NONE, false);
      return true;
    }
    return false;
  ```

Arxalc replied:
Is there a benefit of this code over mine? It seems to just be changing the order of checking conditions in more lines.

Arxalc replied:
Replaced 👍
